File: apps/api/src/schemas/jobspec.ts
import { z } from "zod";

const BackendSchema = z.object({
  type: z.string(),
  model_name: z.string(),
  provider: z.string().optional(),
  base_url: z.string().optional(),
  parallel_calls_count: z.number().int().positive().optional(),
});

const GenerationSchema = z.object({
  temperature: z.number().min(0).max(2).optional(),
  max_new_tokens: z.number().int().positive().optional(),
  top_p: z.number().min(0).max(1).optional(),
});

const TaskEntrySchema = z.object({
  id: z.string(),
  fewshot: z.number().int().nonnegative(),
  params: z.record(z.any()).optional(),
});

const CustomPluginSchema = z.object({
  name: z.string(),
  version: z.string(),
});

const ArtifactsSchema = z.object({
  output_dir: z.string(),
  save_details: z.boolean().default(false),
});

const JudgeEntrySchema = z.object({
  judge_model_name: z.string(),
  judge_backend: z.string(),
  weight: z.number().optional(),
  max_tokens: z.number().int().positive().optional(),
  judge_template_id: z.string().optional(),
});

const ScoringConfigSchema = z.object({
  judge_template_id: z.string().optional(),
  judges: z.array(JudgeEntrySchema).optional(),
  aggregation: z.string().optional(),
  report_disagreement: z.boolean().optional(),
});

export const JobSpecSchema = z.object({
  jobspec_version: z.string().default("1"),
  benchmark_pack_id: z.string(),
  run_name: z.string().optional(),
  max_samples: z.number().int().positive().optional(),
  num_fewshot_seeds: z.number().int().positive().default(1),
  backend: BackendSchema,
  generation: GenerationSchema.optional(),
  tasks: z.array(TaskEntrySchema).nonempty(),
  custom_plugins: z.array(CustomPluginSchema).optional(),
  artifacts: ArtifactsSchema,
  scoring_mode: z.enum(["deterministic", "judge", "jury"]).optional(),
  scoring_config: ScoringConfigSchema.optional(),
});

export type JobSpec = z.infer<typeof JobSpecSchema>;

File: services/runner/runner/jobspec.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class BackendConfig(BaseModel):
  type: str
  model_name: str
  provider: Optional[str] = None
  base_url: Optional[str] = None
  parallel_calls_count: Optional[int] = Field(default=None, gt=0)


class GenerationConfig(BaseModel):
  temperature: Optional[float] = Field(default=None, ge=0.0, le=2.0)
  max_new_tokens: Optional[int] = Field(default=None, gt=0)
  top_p: Optional[float] = Field(default=None, ge=0.0, le=1.0)


class TaskEntry(BaseModel):
  id: str
  fewshot: int = Field(ge=0)
  params: Optional[Dict[str, Any]] = None


class CustomPlugin(BaseModel):
  name: str
  version: str


class ArtifactsConfig(BaseModel):
  output_dir: str
  save_details: bool = False


class JudgeEntry(BaseModel):
  judge_model_name: str
  judge_backend: str
  weight: Optional[float] = None
  max_tokens: Optional[int] = Field(default=None, gt=0)
  judge_template_id: Optional[str] = None


class ScoringConfig(BaseModel):
  judge_template_id: Optional[str] = None
  judges: Optional[List[JudgeEntry]] = None
  aggregation: Optional[str] = None
  report_disagreement: Optional[bool] = None


class JobSpec(BaseModel):
  jobspec_version: str = "1"
  benchmark_pack_id: str
  run_name: Optional[str] = None
  max_samples: Optional[int] = Field(default=None, gt=0)
  num_fewshot_seeds: int = Field(default=1, gt=0)
  backend: BackendConfig
  generation: Optional[GenerationConfig] = None
  tasks: List[TaskEntry]
  custom_plugins: Optional[List[CustomPlugin]] = None
  artifacts: ArtifactsConfig
  scoring_mode: Optional[str] = Field(default=None)
  scoring_config: Optional[ScoringConfig] = None
